precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,615 @@
+	msg := banktypes.NewMsgSend(bridgeAddrBytes, senderAddr.Bytes(), coins)
+
+	if err := m.validateAuthorizationLimits(sendAuth, coins); err != nil {
+		return nil, fmt.Errorf("authorization validation failed: %w", err)
+	}
+
+	_, err = m.authzKeeper.DispatchActions(context.SdkCtx(), bridgeAddr, []sdk.Msg{msg})
+	if err != nil {
+		return nil, err
+	}
+
+	assetUnlocked, err := m.bridgeKeeper.AssetsUnlocked(
+		context.SdkCtx(),
+		inputs.Token.Bytes(),
+		sdkAmount,
+		uint8(inputs.Chain),
+		inputs.Recipient,
+	)
+	if err != nil {
+		// TODO(JEREMY): error happened here, funds might have been
+		// lost, should we panic?
+		return nil, fmt.Errorf("failed to send AssetsUnlocked to bridge: %w", err)
+	}
+
+	// finally update the journal entries to propagate the changes
+	// done to the gas token (BTC in our case)
+	balanceDelta, overflow := uint256.FromBig(inputs.Amount)
+	if overflow {
+		return nil, fmt.Errorf("conversion from big.Int to uint256.Int overflowed: %v", inputs.Amount)
+	}
+
+	// only one side of the transfer to update here as we
+	// burnt funds
+	journal := context.Journal()
+	journal.SubBalance(senderAddr, balanceDelta, tracing.BalanceChangeTransfer)
+
+	return assetUnlocked, nil
+}
+
+func (m *BridgeOutMethod) validateAuthorizationLimits(
+	sendAuth *banktypes.SendAuthorization,
+	requestedCoins sdk.Coins,
+) error {
+	if sendAuth.SpendLimit == nil || sendAuth.SpendLimit.Empty() {
+		return fmt.Errorf("no allowance for for %v", requestedCoins[0].Denom)
+	}
+
+	for _, requestedCoin := range requestedCoins {
+		allowedAmount := sendAuth.SpendLimit.AmountOf(requestedCoin.Denom)
+		if allowedAmount.IsZero() {
+			return fmt.Errorf("no allowance for %s", requestedCoin.Denom)
+		}
+		if requestedCoin.Amount.GT(allowedAmount) {
+			return fmt.Errorf(
+				"requested amount %s exceeds allowed amount %s for %s",
+				requestedCoin.Amount,
+				allowedAmount,
+				requestedCoin.Denom,
+			)
+		}
+	}
+
+	// Check allowed list if it exists
+	// It shouldn't be set seeing that we don't really use cosmos-sdk
+	// but just in case?
+	if len(sendAuth.AllowList) > 0 {
+		found := false
+		senderAddrStr := sdk.AccAddress(common.HexToAddress(
+			evmtypes.AssetsBridgePrecompileAddress,
+		).Bytes()).String()
+
+		for _, allowedAddr := range sendAuth.AllowList {
+			if allowedAddr == senderAddrStr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("recipient address not in authorization allow list")
+		}
+	}
+
+	return nil
+}
+
+// validate applies more extensive validation to the inputs, not only
+// the semantic validation but also verify balances, and tokens etc.
+func (m *BridgeOutMethod) validate(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) error {
+	sdkCtx := context.SdkCtx()
+	// first check that the token is either BTC or a supported ERC20
+	if !m.isValidToken(sdkCtx, inputs.Token) {
+		return fmt.Errorf("unsupported token: %v", inputs.Token)
+	}
+
+	spenderAddr := context.MsgSender()
```

Comment:
The address calling this method is the owner of the balance. The spender is the `AssetsBridge` precompile.

Reply:
fixed in: cab86f65